Docker Guide (#10252)
PrefectHQ/prefect: docs/guides/deployment/docker.md

---

> description: Learn how to store your Prefect code in Docker images and build Prefect deployments that create flow runs in Docker containers.

Reviewer: ```suggestion
description: Learn how to store your flow code in a Docker image and build Prefect deployments that create flow runs in Docker containers.
```

> In this guide, we'll further configure the deployment so flow runs are executed in a Docker container. We'll run our Docker instance locally, but you can extend this guide to run it on remote machines.

Reviewer: ```suggestion
In this guide, we will "bake" your code directly into a Docker image. This will allow for easy storage and allow you to deploy flow code without the need for a storage block. Then, we'll configure the deployment so flow runs are executed in a Docker container based on that image. We'll run our Docker instance locally, but you can extend this guide to run it on remote machines.
```

> This guide will also dive into how you can "bake" your code directly into your Docker image.  This will allow for easy storage and allow you to deploy flow code without the need for a storage block.  

Reviewer: ```suggestion
```

In this guide we'll:

> - Create a Docker image that stores your Prefect flow code

Reviewer: ```suggestion
- Create a Docker image that stores your Prefect flow code.
```

> - Configure a [build step](/concepts/deployments#build) which will build a docker image on our behalf

Reviewer: ```suggestion
- Configure a [build step](/concepts/deployments#build) which will build a docker image on our behalf.
```

## Storing Prefect Flow Code in a Docker Image 

First let's create a directory to work from, `docker-tutorial`.

In this directory, you will create a sub-directory named `flows` and put your flow script from the [Deployments](/tutorial/deployments/) tutorial. In this case, I've named the flow `docker-tutorial-flow.py`.

The next file you will add to the `docker-tutorial` directory is a `requirements.txt`.  In this file make sure to include all dependencies that are required for your `docker-tutorial-flow.py` script.  

Next, you will create a `Dockerfile` that will be used to create a Docker image that will also store the flow code.  This `Dockerfile` should look similar to the following:  

```bash
FROM prefecthq/prefect:2-python3.9
COPY requirements.txt .
RUN pip install -r requirements.txt --trusted-host pypi.python.org --no-cache-dir
ADD flows /opt/prefect/flows
```

Finally, we build this image by running: 

```bash
docker build -t docker-tutorial-image .
```

This will create a Docker image in your Docker repository with your Prefect flow stored in the image.  

## Prefect.yaml File

> Now that you have created a Docker image that stores your Prefect flow code, you're going to make use of Prefect's deployment recipes. In your terminal run:

Reviewer: ```suggestion
Now that you have created a Docker image that stores your Prefect flow code, you're going to make use of Prefect's deployment recipes. In your terminal, run:
```

# control along with your flow code.

# Generic metadata about this project
name: docker-tutorial
prefect-version: 2.10.16

# build section allows you to manage and build docker images
build:
- prefect_docker.deployments.steps.build_docker_image:
    id: build_image
    requires: prefect-docker>=0.3.0
    image_name: docker-tutorial-image
    tag: latest
    dockerfile: auto
    push: true

# push section allows you to manage if and how this project is uploaded to remote locations
push: null

# pull section allows you to provide instructions for cloning this project in remote locations
pull:
- prefect.deployments.steps.set_working_directory:
    directory: /opt/prefect/docker-tutorial

# the deployments section allows you to provide configuration for deploying flows
deployments:
- name: null
  version: null
  tags: []
  description: null
  schedule: {}
  flow_name: null
  entrypoint: null
  parameters: {}
  work_pool:
    name: null
    work_queue_name: null
    job_variables:
      image: '{{ build_image.image }}'
```

Once you have made any updates you would like to this `prefect.yaml` file, in your terminal run: 

<div class="terminal">
```bash
prefect deploy
```
</div>

> The Prefect deployment wizard will walk you through the deployment experience to deploy your flow in either Prefect Cloud or your local Prefect Server.  Once the flow is deployed you can even save the configuration to your `prefect.yaml` file for faster deployment in the future.

Reviewer: ```suggestion
The Prefect deployment wizard will walk you through the deployment experience to deploy your flow in either Prefect Cloud or your local Prefect Server.  Once the flow is deployed, you can even save the configuration to your `prefect.yaml` file for faster deployment in the future.
```

    
    If you shut down the server, you can start it again by opening another terminal session and starting the Prefect server with the `prefect server start` CLI command.

## Storing Prefect Flow Code in a Docker Image 

> First let's create a directory to work from, calling it `docker-tutorial`.

Reviewer: ```suggestion
First let's create a directory to work from, `docker-tutorial`.
```

> In this directory you will create a sub-directory named `flows` and put your flow script from the [Deployments](/tutorial/deployments/) tutorial.  In this case I named the flow `docker-tutorial-flow.py`.

Reviewer: ```suggestion
In this directory, you will create a sub-directory named `flows` and put your flow script from the [Deployments](/tutorial/deployments/) tutorial. In this case, I've named the flow `docker-tutorial-flow.py`.
```

The next file you will add to the `docker-tutorial` directory is a `requirements.txt`.  In this file make sure to include all dependencies that are required for your `docker-tutorial-flow.py` script.  

Next, you will create a `Dockerfile` that will be used to create a Docker image that will also store the flow code.  This `Dockerfile` should look similar to the following:  

```bash
FROM prefecthq/prefect:2-python3.9
COPY requirements.txt .
RUN pip install -r requirements.txt --trusted-host pypi.python.org --no-cache-dir
ADD flows /opt/prefect/flows
```

> Finally you will builds this image by running: 

Reviewer: ```suggestion
Finally, we build this image by running: 
```

```bash
docker build -t docker-tutorial-image .
```

This will create a Docker image in your Docker repository with your Prefect flow stored in the image.  

## Prefect.yaml File

Now that you have created a Docker image that stores your Prefect flow code, you're going to make use of Prefect's deployment recipes. In your terminal run:

<div class="terminal">
```bash
prefect init --recipe docker
```
</div>

You will see a prompt to input values for image name and tag, lets use:
```
image_name: docker-tutorial-image
tag: latest
```

> This will create a `prefect.yaml` file for us populated with some fields, by default it will look like this:

Reviewer: ```suggestion
This will create a `prefect.yaml` file for us populated with some fields. By default, it will look like this:
```
